atspi-connection/src/p2p.rs

```diff
@@ -0,0 +1,746 @@
+									let new_owner_unique_name = OwnedUniqueName::from(new_owner_unique_name.to_owned());
+
+									let _result = self.insert_well_known(well_known_name, new_owner_unique_name, conn).await;
+
+									#[cfg(feature = "tracing")]
+									_result.unwrap_or_else(|err| {
+										warn!("Failed to insert well-known name: {} - {}", well_known_name, err);
+									});
+								}
+
+								// Well-known name appeared on the bus.
+								(Some(old_owner_unique_name), None) => {
+									let well_known_name = OwnedWellKnownName::from(well_known_name);
+									let old_owner_unique_name = OwnedUniqueName::from(old_owner_unique_name.to_owned());
+
+									let _result = self.remove_well_known(&well_known_name, &old_owner_unique_name).await;
+
+									#[cfg(feature = "tracing")]
+									_result.unwrap_or_else(|err| {
+										warn!("Failed to remove well-known name: {} - {}", well_known_name, err);
+									});
+								},
+
+								// Well-known name received a new owner on the bus.
+								(Some(old_owner_unique_name), Some(new_owner_unique_name)) => {
+									let well_known_name = OwnedWellKnownName::from(well_known_name);
+									let old_owner_unique_name = OwnedUniqueName::from(old_owner_unique_name.to_owned());
+									let new_owner_unique_name = OwnedUniqueName::from(new_owner_unique_name.to_owned());
+									let _result = self.update_well_known_owner(
+										well_known_name,
+										old_owner_unique_name,
+										new_owner_unique_name,
+										conn,
+									).await;
+
+									#[cfg(feature = "tracing")]
+									_result.unwrap_or_else(|err| warn!(
+										"Failed to replace well known name {well_known_name} with old owner: {old_owner_unique_name} and new owner: {new_owner_unique_name}: {err}"
+									));
+								}
+							}
+						}
+					}
+				}
+
+				// Handle the case where the stream is closed
+				#[cfg(feature = "tracing")]
+				warn!("Peer listener task stopped, no more events will be processed.");
+				self.clear().await;
+			})
+			.detach();
+	}
+
+	/// Clears the list of peers.
+	///
+	/// # Note
+	/// This is used to reset the list of peers when the D-Bus connection is lost.
+	async fn clear(&self) {
+		let mut peers = self.peers.lock().await;
+		peers.clear();
+	}
+}
+
+impl Peer {
+	/// Creates a new `Peer` with the given bus name and socket path.
+	///
+	/// # Note
+	/// This function is intended for use in building the initial list of peers.
+	///
+	/// If given a `UniqueName`, it will check if the peer also owns a well-known name.
+	/// If given a `WellKnownName`, it will query the D-Bus for the unique name of the peer.
+	///
+	/// # Errors
+	/// - `DBusProxy` cannot be created.
+	/// - The socket address cannot be parsed.
+	///
+	pub(crate) async fn try_new<B, S>(
+		bus_name: B,
+		socket: S,
+		conn: &zbus::Connection,
+	) -> Result<Self, AtspiError>
+	where
+		B: Into<OwnedBusName>,
+		S: TryInto<Address>,
```

Comment:
Solid trait bounds here! This is much more useful than taking direct parameters.